File: simpleburner.pl
#!/usr/bin/env perl
# file: made to simplify CD/DVD burning under CLI
#  
#   Copyright 2009, 2010 Marcin Karpezo <sirmacik at gmail dot com>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

use encoding 'utf8';

use strict;
use warnings;
use Getopt::Long;

my ($writer, $isomaker, $speed, $test, $burn, $make);
my $datadir = "/tmp";
my $isoname = "~/cd.iso";
my $device = "/dev/sr0";
my $mode = "tao";

GetOptions("data=s" => \$datadir,
           "name=s" => \$isoname,
           "device=s" => \$device,
           "speed=s" => \$speed,
           "mode=s" => \$mode,
           "t|test" => \$test,
           "b|burn-only" => \$burn,
           "m|makeiso" => \$make,
           "h|help" => \&helpmsg,);

sub warning {
    print <<END;
simpleburner, made to simplify CD/DVD burning under CLI

simpleburner Copyright (C) 2009, 2010 Marcin Karpezo 
This program comes with ABSOLUTELY NO WARRANTY. 
This is free software, and you are welcome 
to redistribute it under certain conditions.
For details see COPYING file.

END
}

sub helpmsg {
    warning();
    print <<EOM;
Simpleburner, made to simplify CD/DVD burning under CLI
Usage: simpleburner [options]    
    
    -h, --help      Displays this message
        --data      Directory with data to burn
        --name      Path and/or name of iso image (by default ~/cd.iso)
        --device    Device to use (default /dev/sr0)
        --speed     Burning speed (by default it will be autodetected)
        --mode      Burning mode, available options are: TAO (default), DAO, SAO, RAW
    -t, --test      Run in test mode
    -b, --burn-only Run without making iso image
    -m, --makeiso   Make only iso image

Please send any bug reports or feuture requests to simpleburner-bugs\@googlegroups.com
EOM
    exit 0;
}

sub try {
    print shift;
    shift();
    print shift || "[OK!]\n";
}

sub programcheck {
    try "Looking for cdrkit...", do {
        if (-e '/usr/bin/wodim' and -e '/usr/bin/genisoimage') {
            $writer = 'wodim';
            $isomaker = 'genisoimage';
        } elsif (-e '/usr/bin/cdrecord' and -e '/usr/bin/mkisofs') {
            print("Not found\nLooking for cdrtools...");
            $writer = 'cdrecord';
            $isomaker = 'mkisofs';
        } else {
            die("Not found: Please install cdrkit or cdrtools!\n");
        }
    };
}

sub optcheck {
    unless ($burn or $datadir) {
        die("Failed! You must define --data option.\n"
            ."Run -h|--help for more information.\n");
    } 
    unless (-d $datadir) {
        die("Failed! Data directory '$datadir' does not exist.\n");
    }
}

sub makeiso {
    if (-e $isoname) {
        oldisock();
    }
    try "Making iso image", do {
        $datadir =~ s/\s+/\\ /g;
        my $command = "$isomaker -UR -quiet -allow-multidot "
                      ."-allow-leading-dots -iso-level 3 "
                      ."-o $isoname $datadir";
        system($command) or die "Can't make iso file!\n";
    };
    print("File stored in $isoname\n");
}

sub burniso {
    try "Burning iso...", do {
        my ($burnspeed, $runtest);
        if ($speed) {
            $burnspeed = " --speed=$speed";
        } elsif ($test) {
            $runtest = "--dummy";
        }
        my $command = "$writer --eject -vs -$mode --dev=$device "
                     ."$burnspeed $runtest $isoname";
        system("$command > /dev/null") or die "Can't burn disc!\n";
    };
}

sub oldisock {
    print("Old iso file detected, delete it? [Y/n] ");
    my $reply = <STDIN>; chomp $reply;
    if ($reply eq "n") {
        print("Burn it? [Y/n] ");
        my $reply = <STDIN>; chomp $reply;
        if ($reply eq "n") {
            die("Stopped!\n");
        } else {
            burniso();
            exit 0;
        }
    } else {
        try "Deleting old iso file...", { unlink $isoname };
        makeiso();
        burniso();
    }
}

$isoname =~ s/^~/$ENV{'HOME'}/;
$datadir =~ s/^~/$ENV{'HOME'}/;

warning();
programcheck();
if ($burn) {
    burniso();
} elsif ($make) {
    makeiso();
    optcheck();
} else {
    optcheck();
    oldisock();
}

#TODO:
# * Delete of iso image after burning, nie, 3 sty 2010, 19:04:50 CET
# * CD-RW cleaning, nie, 3 sty 2010, 19:05:14 CET
# ** Add audiocd burning 2009-10-26 21:33+0100

File: simpleburner.py
#!/usr/bin/env python
#-*- coding: utf-8 -*-
# Copyright (c) 2009, Marcin Karpezo 
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification, 
# are permitted provided that the following conditions are met:
#
#     * Redistributions of source code must retain the above copyright notice, 
#       this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright notice, 
#       this list of conditions and the following disclaimer in the documentation 
#       and/or other materials provided with the distribution.
#     * Neither the name of the simpleburner nor the names of its contributors may 
#       be used to endorse or promote products derived from this software without 
#       specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND 
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED 
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. 
# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, 
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, 
# OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
# WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.

import os
import sys
import subprocess
from optparse import OptionParser

def programcheck():
    print "Looking for cdrkit..."
    if os.path.exists('/usr/bin/wodim') and os.path.exists('/usr/bin/genisoimage'):
        globals() ["writer"] = 'wodim'
        globals() ["isomaker"] = 'genisoimage'
        print "[OK]"

    else:
        print >>sys.stderr, "Not found\nLooking for cdrtools..."
        if os.path.exists('/usr/bin/cdrecord') and os.path.exists('/usr/bin/mkisofs'):
            globals() ["writer"] = 'cdrecord'
            globals() ["isomaker"] = 'mkisofs'
            print "[OK]"
        else:
            print >>sys.stderr, "Not found: Please install cdrkit or cdrtools!"
            sys.exit(1)
def optcheck():
    if not burn:
        if not datadir:
            print >>sys.stderr, "Failed! You must deine --data option."
            sys.exit(1)
        elif not os.path.exists(datadir):
            print >>sys.stderr, "Failed! Data directory does not exist!"
            sys.exit(1)
def makeiso():
    print "Making iso image..."
    command = "%s -U -quiet -o %s %s" % (isomaker, isoname, datadir)
    makeiso = subprocess.Popen(command, stdout=subprocess.PIPE, shell=True).communicate()[0]
    if makeiso:
        print >>sys.stderr, "Failed!"
        sys.exit(1)
    else:
        print "[OK]\nFile stored in %s" % isoname
def burniso():
    print "Burning iso..."
    if speed:
        burnspeed = " --speed=%s" % speed
    else:
        burnspeed = ''
    if test == "True":
        runtest = ' --dummy'
    else:
        runtest = ''
    command = "%s --eject -vs -%s --dev=%s %s %s %s" % (writer, mode, device, burnspeed, runtest, isoname)
    burniso = subprocess.Popen(command, stdout=subprocess.PIPE, shell=True).communicate()[0]
    if burniso:
        print >>sys.stderr, "Failed or ended successfully with warnings"
        sys.exit(1)
    else:
        print "[OK]"

if __name__=="__main__":
    usage = "usage: %prog [options]"
    parser = OptionParser(usage=usage)
    parser.add_option("--data", dest="datadir",
                      help="set directory with data to burn")
    parser.add_option("--name", dest="isoname", default="/tmp/cd.iso",
                      help="set path and/or name of iso image (by default /tmp/cd.iso)")
    parser.add_option("--device", dest="device", default="/dev/sr0",
                      help="set device to use (default /dev/sr0)")
    parser.add_option("--speed", dest="speed",
                      help="set burn speed (by default it will be autodetected)")
    parser.add_option("--mode", dest="mode", default="tao",
                      help="set burn mode; available options are: TAO (default), DAO, SAO, RAW)")
    parser.add_option("-t", "--test", action="store_true", dest="test", default= False ,
                      help="run in test burn mode")
    parser.add_option("-b", "--burn-only", action="store_true", dest="burn", default=False,
                      help="run without making iso image")
    parser.add_option("-m", "--makeiso", action="store_true", dest="make", default=False,
                      help="make only iso image")

    (options, args) = parser.parse_args()
    datadir = options.datadir
    isoname = options.isoname
    device = options.device
    speed = options.speed
    mode = options.mode
    test = options.test
    burn = options.burn
    make = options.make

    programcheck()
    optcheck()
    if burn == True:
        burniso()
    elif make == True:
        makeiso()
    else:
        makeiso()
        burniso()
        
